### PR TITLE
[DPTP-2094] ci-secret-bootstrap: validate if the given items/field/attachments exist

### PR DESCRIPTION
**Pull request by droslean**
/cc @openshift/openshift-team-developer-productivity-test-platform 

Signed-off-by: Nikolaos Moraitis <nmoraiti@redhat.com>

### CODE REVIEW

**Review by stevekuznetsov — COMMENTED** [review]
Technically this is an avenue for attack (to list items in BW) if we run in a presubmit but I think it's useful enough to have?

**Comment by droslean** [comment]
> Technically this is an avenue for attack (to list items in BW) if we run in a presubmit but I think it's useful enough to have?

Well, yes. If someone opens a PR adding a trillion secrets in the config, we will brute force Bitwarder, but I think this is fine since `ok-to-test` is protecting us. My main concern is if we still care about Bitwarden, since we are planning to migrate our secrets to vault.

File: cmd/ci-secret-bootstrap/main.go [hunk 1]
```diff
@@ -672,23 +672,63 @@ func getUnusedBWItems(config secretbootstrap.Config, client secrets.Client, bwAl
...
+			} else {
+				if !client.HasItem(item.BWItem) {
+					errs = append(errs, fmt.Errorf("item %s doesn't exist", item.BWItem))
+					continue
+				}
+
+				if item.Field != "" {
+					if _, err := client.GetFieldOnItem(item.BWItem, item.Field); err != nil {
+						errs = append(errs, fmt.Errorf("field %s in item %s doesn't exist", item.Field, item.BWItem))
+						continue
```
Comment (petr-muller) [inline cmd/ci-secret-bootstrap/main.go:701]: These `continue`s` (this one and the ones below) do not make sense to me. If we expect exactly one of the `if`s to apply, then the three `if` statements should be a switch, or we expect multiple ones to apply, and then we should not abort on finding a first error.

Comment (droslean) [reply cmd/ci-secret-bootstrap/main.go:701]: Nice catch. I will update it. Thanks

Comment (droslean) [reply cmd/ci-secret-bootstrap/main.go]: updated

File: cmd/ci-secret-bootstrap/main.go [hunk 2]
```diff
@@ -741,23 +741,59 @@ func getUnusedBWItems(config secretbootstrap.Config, client secrets.ReadOnlyClie
...
+	var errs []error
+
+	for _, config := range secretConfigs {
+		for _, item := range config.From {
+
+			if item.DockerConfigJSONData != nil {
+				for _, data := range item.DockerConfigJSONData {
+					if !client.HasItem(data.BWItem) {
+						errs = append(errs, fmt.Errorf("item %s doesn't exist", data.BWItem))
+						continue
```
Comment (stevekuznetsov) [inline cmd/ci-secret-bootstrap/main.go:754]: This `continue` does not make sense - the next check is for an attachment to this item, so if the item does not exist there's no point in checking for the attachment.

Comment (droslean) [reply cmd/ci-secret-bootstrap/main.go:754]: :facepalm:  Let me fix this

Comment (droslean) [reply cmd/ci-secret-bootstrap/main.go]: updated

File: pkg/secrets/vault.go [hunk 3]
```diff
@@ -75,6 +75,12 @@ func (c *vaultClient) setItemAtPath(path, field string, content string) error {
 	return c.upstream.UpsertKV(path, data)
 }
 
+func (c *vaultClient) HasItem(itemName string) bool {
+	path := c.pathFor(itemName)
+	_, err := c.upstream.GetKV(path)
+	return err != nil
```
Comment (alvaroaleman) [inline pkg/secrets/vault.go:81]: Please make this a IsNotFound check instead

Comment (droslean) [reply pkg/secrets/vault.go]: updated.

File: pkg/secrets/vault.go [hunk 4]
```diff
@@ -75,6 +75,12 @@ func (c *vaultClient) setItemAtPath(path, field string, content string) error {
 	return c.upstream.UpsertKV(path, data)
 }
 
+func (c *vaultClient) HasItem(itemName string) bool {
+	path := c.pathFor(itemName)
+	_, err := c.upstream.GetKV(path)
+	return vaultclient.IsNotFound(err)
```
Comment (alvaroaleman) [inline pkg/secrets/vault.go:81]: THis needs to be negated, if it is a notfound, you need to return false, not true

Comment (droslean) [reply pkg/secrets/vault.go]: :man_facepalming:

**Comment by alvaroaleman** [comment]
Gimme link for DPTP-2094

**Review by alvaroaleman — APPROVED** [review]

**Pull request merged**